include/envoy/network/transport_socket.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/pure.h"
+#include "envoy/network/connection.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Action that should occur on a connection after I/O.
+ */
+enum class PostIoAction {
+  // Close the connection.
+  Close,
+  // Keep the connection open.
+  KeepOpen
+};
+
+/**
+ * Result of each I/O event.
+ */
+struct IoResult {
+  PostIoAction action_;
+
+  /**
+   * Number of bytes processed by the I/O event.
+   */
+  uint64_t bytes_processed_;
+};
+
+/**
+ * Callbacks used by transport socket instances to communicate with connection
+ */
+class TransportSocketCallbacks {
+public:
+  virtual ~TransportSocketCallbacks() {}
+
+  /**
+   * @return int the file descriptor associated with the connection.
+   */
+  virtual int fd() PURE;
```

Comment:
Not all transport sockets are going to have a fd. Can this be removed from the interface somehow? (I haven't yet looked in detail at how the rest of the code is factored).

In now reading more of the PR, I guess this is due to the fact that fundamentally we are passing FDs around from listeners. I suppose at some point we may need to turn this into some type of generic handle but this is probably ok for now. Just thinking out loud.

Reply:
Yeah I thought we need some type of generic handle but it is not easy and out of scope of this PR. We can refactor when we have something other than FDs.

Reply:
I had the same concern but the SSL stack at least needs this.

We never found a good workaround in the GFE - just made -1 a permissible value which allows all sorts of bad feature creep :-(